### DISCUSSION TITLE
Pre-Release Questions

### DISCUSSION THREAD

**Discussion by cyberbit** [cyberbit/telem#12] [Q&A]
In these very early days of Telem, documentation is scarce. If you have questions on how to get something set up, please ask here!

**Comment by DinnerBeef** [comment]
Trying to get fluid working with this. What Type of tanks are supported. Trying with a dynamic tank, and it keeps saying "attempt to call filed  "tanks" (a nil value)

**Reply by cyberbit** [reply]
Most fluid tanks should work with the fluid storage adapter, as it uses the standard API for fluids. However, Mekanism has done a lot customizations with regards to how their structures interact with ComputerCraft, so it looks like a `telem.input.mekanism.dynamicTank` adapter will be required. Same with Chemical Tanks and Fluid Tanks from Mekanism as well, unfortunately....

In the meantime, if you are on `0.1.0` or higher, you can use the custom adapter definition below until I can integrate full support.

Thanks for the question!

```lua
local telem = require 'telem'

-- define custom adapter
local inputMekanismDynamicTank = function (source)
    return function ()
        local dynamicTank = peripheral.wrap(source)
        local prefix = 'mekdyntank:'

        local stored = dynamicTank.getStored()

        return
            -- machine metrics
            telem.metric{ name = prefix .. 'chemical_tank_capacity', value = dynamicTank.getChemicalTankCapacity() / 1000, unit = 'B', source = source },
            telem.metric{ name = prefix .. 'filled_percentage', value = dynamicTank.getFilledPercentage(), unit = nil, source = source },
            telem.metric{ name = prefix .. 'tank_capacity', value = dynamicTank.getTankCapacity() / 1000, unit = 'B', source = source },

            -- storage metrics
            telem.metric{ name = 'storage:' .. stored.name, value = stored.amount / 1000, unit = 'B', source = source }
    end
end

local backplane = telem.backplane()
    :addInput('tank_0', telem.input.custom(inputMekanismDynamicTank('dynamicValve_0')))
    :addInput('tank_1', telem.input.custom(inputMekanismDynamicTank('dynamicValve_1')))
    :addOutput('hout', telem.output.helloWorld())

backplane:cycleEvery(3)()
```

![image](https://github.com/cyberbit/telem/assets/7350183/02edd0a6-4be4-4aeb-8515-e0c31cb89374)

**Comment by DeveloperNeon** [comment]
Heya!

I was looking at the gallery, and one of the grafana graphs is exactly what I need. Would you be able to share the code you used for that? Im looking at the reactor and induction matrix monitoring grafana dashboard (image below).

Thanks!

![image](https://github.com/cyberbit/telem/assets/51766327/45273737-2bc4-4526-9f7e-869db6a45d2d)

**Reply by cyberbit** [reply]
Hello!

**If you are using Grafana Cloud**, a good starting point would be the JSON export of this panel I made a little while ago. It is at the bottom of API / GrafanaOutputAdapter in the docs: https://telem.notion.site/GrafanaOutputAdapter-3a0b62c505eb44e7aedaf2a33be0f8bf. You should be able to import using New > Import on the Dashboards page.

![image](https://github.com/cyberbit/telem/assets/7350183/df2fa628-0077-455d-bcb1-0ba3e8fa3c7a)

I have been lacking time to write a comprehensive Grafana Cloud guide, so apologies in advance! Here are some things to note:
- Buffer Energy (unrelated to turbine energy) is broken because I was using a prototype adapter for a non-Mekanism energy meter at the time, which is not yet released in Telem.
- Fusion Production is "broken" in versions of Minecraft before 1.20.x because of a Mekanism API bug. The current workaround is to use a CustomInputAdapter to read an energy meter between the fusion output and your power circuit (Energy Meter mod or Energy Detector from Advanced Peripherals work well).
- Label filters and Operations on the panel queries may need to be re-sourced for your particular setup.
- Bar Gauges use Overrides for the color thresholds. Tweak them to whatever makes sense to you.
- Gauges use two queries and Transform data settings to dynamically assign a maximum.

**If you are using self-hosted Grafana**, there is fairly detailed information on the page linked above how to set up Grafana and some basic guidance on the InfluxDB query builder, but I have not yet built out the Mekanism dashboard using that methodology. In theory you could import the dashboard JSON, but I have not tested this and you will probably have to create new queries anyway.

Let me know if this is what you are looking for, and please don't hesitate to reach out if you have issues or more questions!

**Reply by DeveloperNeon** [reply]
Alright, thanks!

I apparently missed that part of the docs ha ha. Anyways, hopefully this weekend Ill get a chance to play with the dash.

**Reply by DeveloperNeon** [reply]
Would you be able to provide me with the code I would put on the computer that is sending the data to grafana, or some sort of example that integrates with grafana?

**Reply by cyberbit** [reply]
Apologies, I thought I had a code example in the docs but it looks like I do not! I will get that added in the next update.

Please see code below for an example Telem script that transmits Fission Reactor and Industrial Turbine metrics to Grafana Cloud. You will need to visit https://instancename.grafana.net/connections/add-new-connection/http-metrics (replace `instancename` with your instance name) and follow the steps to generate your API key and endpoint at the end of Step 4:

![image](https://github.com/cyberbit/telem/assets/7350183/27885c28-7876-48ac-ac27-dd03925ab617)


```lua
local telem = require 'telem'

local authGrafana = {
  endpoint = 'https://influx-prod-xyz.grafana.net/api/v1/push/influx/write',
  apiKey = '10xxx:glc_eyzzz'
}

local backplane = telem.backplane()
  -- inputs
  :addInput('fission', telem.input.mekanism.fissionReactor('fissionReactorLogicAdapter_1'))
  :addInput('turbine', telem.input.mekanism.industrialTurbine('turbineValve_1'))

  -- outputs
  :addOutput('grafana', telem.output.grafana(authGrafana.endpoint, authGrafana.apiKey))

-- push to Grafana every 30 seconds
parallel.waitForAny(
  backplane:cycleEvery(30)
)
```

That code should push everything you need to set up these panels:

![image](https://github.com/cyberbit/telem/assets/7350183/54958bdd-c001-4602-b207-c65453ae519e)

**Comment by DinnerBeef** [comment]
I have this code running 

~~~
local telem = require "telem"
local basalt = require "basalt"

local monitor = basalt.addMonitor():setMonitor("right")

local backplane = telem.backplane()
    -- inputs
    :addInput("my_induction", telem.input.mekanism.inductionMatrix("inductionPort_0"))
    -- MonitorFrame
    :addOutput("monitor_rand1",
      telem.output.basalt.label(monitor, "mekinduction:energy_input", colors.green, colors.white))
    :addOutput("hout", telem.output.helloWorld())
-- you could use a Basalt thread to run the backplane as well
parallel.waitForAny(
  basalt.autoUpdate,

  -- include cycleEvery last!
  backplane:cycleEvery(1)
)
~~~
However, the output display appears to be blank and I Don't know why.
Doing a Hello World output does show there is data though.

![image](https://github.com/cyberbit/telem/assets/8346986/9dba5567-260a-491e-afcd-9f125bd6553e)
![image](https://github.com/cyberbit/telem/assets/8346986/bf6fa988-401b-4738-886f-1ff166b12bbf)
I even tried doing the random one and nothing is happening

**Reply by cyberbit** [reply]
Looking at this now. What version of Telem do you have?

**Reply by DinnerBeef** [reply]
I just did "wget run https://pinestore.cc/d/14" so should be v0.5.0

**Reply by cyberbit** [reply]
That code seems to be working for me in an emulator when swapping out the induction input for a hello world input, unfortunately not able to test in-game at the moment.

Are there other programs/code running on that computer besides your example? I am aware of a strange bug that affects Basalt outputs that I have not been able to identify, other than it has something to do with the event queue.

**Reply by DinnerBeef** [reply]
Nope i am running just that

**Reply by cyberbit** [reply]
If you comment out the induction input and insert `:addInput('hello', telem.input.helloWorld(123))`, and change the label output metric to `"hello_world"`, what happens

**Reply by DinnerBeef** [reply]
doing the same thing for me "IE still broke"

**Reply by cyberbit** [reply]
That is strange, your code seems fine. I'll have to test this in-game later, apologies!

If you are able to send me this small test world, that would help with debugging on my end.

**Reply by cyberbit** [reply]
I was able to test in a local world. Your code is working for me:

![image](https://github.com/cyberbit/telem/assets/7350183/a90eb387-aa4a-4ed6-9b61-7e06abcf1ed3)

Can you send a list of the mods and their versions in this test pack?

**Reply by DinnerBeef** [reply]
I added a bunch of mods since then. Forget to mention I am on 1.20.4. However, I am pretty sure it was JEI/CC: Tweaked/ Mek

**Reply by cyberbit** [reply]
Wish I had a better answer, but I just pulled 1.20.4 with JEI and Mekanism and it seems to be working for me:

![image](https://github.com/cyberbit/telem/assets/7350183/cffefe4c-f5ce-4241-9607-59b7925035c6)

**Comment by cyberbit** [comment]
Closing this discussion as the documentation site is now considered stable.
